src/Expr.h

```diff
@@ -188,8 +188,11 @@ struct IntImm : public ExprNode<IntImm> {
         internal_assert(t.bits() == 8 || t.bits() == 16 || t.bits() == 32 || t.bits() == 64)
             << "IntImm must be 8, 16, 32, or 64-bit\n";
 
-        // Normalize the value by dropping the high bits
-        value <<= (64 - t.bits());
+        // Normalize the value by dropping the high bits.
+        // Since left-shift of negative value is UB in C++, cast to uint64 first;
+        // it's unlikely any compilers we care about will misbehave, but UBSan will complain.
+        value = (int64_t) (((uint64_t) value) << (64 - t.bits()));
```

Comment:
To be clear - there's no way (64 - t.bits()) can be negative, right? Is UBSan actually encountering a negative value here, or just complaining that it's a shift by a signed amount?

If that's the case, it might be clearer to cast the shift amount to unsigned, rather than the value, e.g:

value <<= (unsigned)(64 - t.bits());

Reply:
I interpreted this to mean 'value is negative'

Reply:
OK, the specific issue is that shifting a 1 bit into or past the sign bit is (technically) an undefined behavior. (Since uints don't have sign bits, they don't have this issue.)

Reply:
Oh ok, I misunderstood